File: TypeScript/rational-numbers.ts
function gcd(a: number, b: number): number {
    while (b !== 0) {
        const temp = b;
        b = a % b;
        a = temp;
    }

    return Math.floor(a);
}

export class Rational {
    numerator: number;
    denominator: number;

    constructor(a: number, b: number) {
        this.numerator = a;
        this.denominator = b;
    }
  
    add(other: Rational): Rational {
        const a = (this.numerator * other.denominator) + (this.denominator * other.numerator);
        const b = this.denominator * other.denominator;
        return new Rational(a, b).reduce();
    }

    sub(other: Rational): Rational {
        let a = this.numerator * other.denominator - this.denominator * other.numerator;
        const b = this.denominator * other.denominator;
        return new Rational(a, b).reduce();
    }

    mul(other: Rational): Rational {
        if(other.numerator === 1 && other.denominator == 1) {
          return this;
        }
        else if(other.numerator === 0) {
          return other;
        }

        let a = (this.numerator * this.denominator);
        let b = other.numerator * other.denominator;
        return new Rational(a, b).reduce();
    }

    div(other: Rational): Rational {
        if(this.denominator === 0) {
            throw new Error('Divide by zero')
        }

        let a = (this.numerator * other.denominator);
        let b = this.denominator * other.numerator;
        return new Rational(a, b).reduce();
    }

    abs(other: Rational): Rational {
        let r: Rational = new Rational(Math.abs(this.numerator), Math.abs(this.denominator));
        return r.reduce();
    }


    exprational(n: number): Rational {
        let r: Rational;

        if (n > 0) {
            r = new Rational(this.numerator ** n, this.denominator ** n);
        }
        else{
            n = Math.abs(n);
            r = new Rational(this.denominator ** n, this.numerator ** n);
        }
        
        return r.reduce();
    }

    expreal(n: number) : number {
        const base = Math.pow(n, 1 / this.denominator);
        return Math.pow(base, this.numerator);
    }

    reduce(): Rational {
        if (this.numerator === 0) {
            return new Rational(0, 1);
        }

        const n = gcd(this.numerator, this.denominator);
        let num = this.numerator / n;
        let denom = this.denominator / n;

        if (denom < 0) {
            num *= -1;
            denom *= -1;
        }

        return new Rational(num, denom);
    }
}

File: C#/Rational Numbers/RationalNumbers.cs
using System;

public static class MathHelper
{
    public static int GCD(int a, int b)
    {
        while(b != 0)
        {
            int tmp = b;
            b = a % b; 
            a = tmp;
        }

        return a;
    }
}

public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r)
    {
        return r.Expreal(realNumber);
    }
}

public struct RationalNumber
{
    public readonly int Numerator = 1;
    public readonly int Denominator = 1;

    public RationalNumber(int numerator, int denominator)
    {
        Numerator = numerator;
        Denominator = denominator;
    }

    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2)
    {
        int a = (r1.Numerator * r2.Denominator) + (r1.Denominator * r2.Numerator);
        int b = r1.Denominator * r2.Denominator;
        return new RationalNumber(a, b).Reduce();
    }

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2)
    {
        int a = (r1.Numerator * r2.Denominator) - (r1.Denominator * r2.Numerator);
        int b = r1.Denominator * r2.Denominator;
        return new RationalNumber(a, b).Reduce();
    }

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2)
    {
        if (r2.Numerator == 1 && r2.Denominator == 1)
        {
            return r1;
        }
        else if (r2.Numerator == 0)
        {
            return r2;
        }

        int a = (r1.Numerator * r1.Denominator);
        int b = r2.Numerator * r2.Denominator;
        return new RationalNumber(a, b).Reduce();
    }

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2)
    {
        if (r1.Denominator == 0)
        {
            throw new DivideByZeroException();
        }

        int a = (r1.Numerator * r2.Denominator);
        int b = r1.Denominator * r2.Numerator;
        return new RationalNumber(a, b).Reduce();
    }

    public RationalNumber Abs()
    {
        RationalNumber rn = new RationalNumber(Math.Abs(this.Numerator), Math.Abs(this.Denominator));
        return rn.Reduce();
    }

    public RationalNumber Reduce()
    {
        if (this.Numerator == 0)
        {
            return new RationalNumber(0, 1);
        }

        int n = MathHelper.GCD(this.Numerator, this.Denominator);
        int num = this.Numerator / n;
        int denom = this.Denominator / n;

        if (denom < 0)
        {
            num *= -1;
            denom *= -1;
        }

        return new RationalNumber(num, denom);
    }

    public RationalNumber Exprational(int power)
    {
        RationalNumber rn = new RationalNumber();

        if(power > 0)
        {
            rn = new RationalNumber((int)Math.Pow(this.Numerator, power), (int)Math.Pow(this.Denominator, power));
            return rn.Reduce();
        }

        int n = Math.Abs(power);
        rn = new RationalNumber((int)Math.Pow(this.Denominator, n), (int)Math.Pow(this.Numerator, n));

        return rn.Reduce();
    }

    public double Expreal(int baseNumber)
    {
        double b = Math.Pow(baseNumber, 1d / this.Denominator);
        return Math.Pow(b, this.Numerator);
    }

    public override string ToString()
    {
        return $"RationalNumber({this.Numerator}, {this.Denominator})";
    }
}